### PR TITLE
Provider stats: add injected live snapshot publisher boundary (closes #1598)

### LINKED ISSUE #1598: Provider stats: add injected live snapshot publisher boundary
## Background

Provider sessions currently increment internal send/receive counters, and `WorkerRegistry.publish_provider_snapshot(repo_name)` copies those counters into `FidoState`. There is no provider-session-side publication boundary for counter mutations.

## Invariant

Provider/session code may request a provider-stats snapshot publish only through an injected collaborator/callback. Provider modules must not import or reach through to `WorkerRegistry`, and status serialization must continue to read only immutable `FidoState` snapshots.

## Scope

- Add a small typed collaborator/protocol/callback for live provider stat publication.
- Wire the collaborator through provider/session construction for Claude, Codex, and Copilot CLI.
- Keep module-level code thin and dependency-injected.
- Do not change send/receive counter publication behavior yet except where needed to prove the boundary is wired.

## Acceptance Criteria

- Unit tests can inject a fake publisher and observe that sessions have the publication boundary available.
- `server.py` status serialization still reads `ProviderSnapshot` from `FidoState`; it does not read provider/session objects directly.
- Provider session modules do not import `WorkerRegistry`.
- No compatibility shim or duplicate legacy path is added.

### PR DESCRIPTION
**Pull request by FidoCanCode**
Fixes #1598.

Extracts `FidoState` and its associated dataclasses (`ProviderSnapshot`, `RepoState`, `ThreadSnapshot`, etc.) from `registry.py` into a new `fido.appstate` leaf module, then injects `AtomicUpdater[FidoState]` through `SessionBackedAgent.__init__` and `DefaultProviderFactory.create_provider`. All three provider agents (Claude, Codex, CopilotCLI) import directly from `fido.atomic` and `fido.appstate` with no back-reference to the registry — no `TYPE_CHECKING` workarounds needed. The composition root calls `create_atomic` inline rather than wrapping it in a helper.

Fixes #1598.

---

## Work queue

<details><summary>Completed (6)</summary>

- [x] [Remove create_fido_atomic() and _EMPTY_FIDO_STATE from appstate.py. Remove the 'create_atomic as _create_atomic' alias import (and the AtomicUpdater import used only for the return annotation) from appstate.py. In server.py, inline the call as create_atomic(FidoState(repos=frozendict(), github_limits=GitHubLimit())). Update all tests that imported create_fido_atomic from fido.appstate to import create_atomic from fido.atomic instead.](https://github.com/FidoCanCode/home/pull/1602#discussion_r3219374198) 
- [x] [Extract FidoState and all its associated dataclasses (ProviderSnapshot, RepoState, ThreadSnapshot, etc.) from registry.py into a new fido.state module. Update registry.py and all other importers accordingly. Provider modules (claude.py, codex.py, copilotcli.py, session_agent.py, provider_factory.py) then import AtomicUpdater from fido.atomic and FidoState from fido.state directly, eliminating the TYPE_CHECKING workarounds.](https://github.com/FidoCanCode/home/pull/1602#discussion_r3219088959) 
- [x] [Replace ProviderStatsPublisher Protocol with AtomicUpdater[FidoState] passed through the constructor chain. Use TYPE_CHECKING import for FidoState in provider modules. Remove NullProviderStatsPublisher.](https://github.com/FidoCanCode/home/pull/1602#issuecomment-4417262023) 
- [x] Inject publisher into SessionBackedAgent and provider factory 
- [x] Add unit tests for the injected publisher boundary 
- [x] Define ProviderStatsPublisher protocol in provider.py 
</details>

### CODE REVIEW

**Comment by rhencke** [comment]
Why do we need a whole new stats publishing class when we already have AtomicWriter[FidoState]?

**Comment by FidoCanCode** [comment]
Good question. The short answer: `AtomicUpdater[FidoState]` knows the shape of `FidoState` and lives in `registry.py`. If provider modules accepted an `AtomicUpdater` directly, they'd need to import `FidoState`, `ProviderSnapshot`, and the lens path into the state tree — which is exactly the coupling #1598 exists to break.

`ProviderStatsPublisher` is a one-method Protocol: `publish() -> None`. The provider says "my stats changed, please snapshot them." It doesn't know what `FidoState` looks like or that `AtomicUpdater` exists. The factory closes over the real updater logic when it wires the callback — provider modules stay ignorant of the registry layer.

So it's not replacing `AtomicUpdater` — it's a thin boundary that keeps the dependency arrow pointing the right way.

**Comment by rhencke** [comment]
From the issue:

"Provider/session code may request a provider-stats snapshot publish only through an injected collaborator/callback. Provider modules must not import or reach through to WorkerRegistry, and status serialization must continue to read only immutable FidoState snapshots."

This is literally what AtomicWriter was made for and we have been passing in everywhere else DI style. It *is* the injected collaborator/callback. Remember, just because you *import* FidoState doesn't mean you are creating one nor accessing one passed in.  If you are really really concerned, just import FidoState under TYPE_CHECKING.

The coupling we are trying to break is the *update* of stats from the *reading* of stats.  It is completely okay for all these classes to know the *shape* of FidoState, it's just not okay for them to *read* or *instantiate* from one (which AtomicWriter guarantees)

**Comment by FidoCanCode** [comment]
Yeah, you're right. I was treating shape-knowledge as the coupling to break, but the actual boundary is update-vs-read — and `AtomicWriter` already enforces that. A second Protocol on top just adds indirection without buying anything.

I'll rework this to pass `AtomicUpdater[FidoState]` through the constructor chain instead of `ProviderStatsPublisher`. Same DI threading, less ceremony.

File: src/fido/claude.py
Comment (rhencke) [inline src/fido/claude.py:1]: https://github.com/FidoCanCode/home/pull/1602#issuecomment-4417266791 please

**Review by rhencke — CHANGES_REQUESTED** [review]

**Comment by FidoCanCode** [comment]
Yeah, you're right. I was treating shape-knowledge as the coupling to break, but the actual boundary is update-vs-read — and `AtomicWriter` already enforces that. A second Protocol on top just adds indirection without buying anything.

I'll rework this to pass `AtomicUpdater[FidoState]` through the constructor chain instead of `ProviderStatsPublisher`. Same DI threading, less ceremony.

File: src/fido/claude.py
Comment (FidoCanCode) [reply src/fido/claude.py:1]: Saw this and tried to respond a few times, but I kept tripping over the structured response format. The rework to `AtomicUpdater[FidoState]` is already queued — I'll get to it on the next worker turn.

**Comment by FidoCanCode** [comment]
Done — swapped out `ProviderStatsPublisher` and `NullProviderStatsPublisher` for `AtomicUpdater[FidoState]` threaded through `SessionBackedAgent`, `ClaudeClient`, `CodexClient`, `CopilotCLIClient`, and `DefaultProviderFactory.create_provider`. `FidoState` imported under `TYPE_CHECKING` in the provider modules. Tests updated to inject a mock `AtomicUpdater` and verify `state_updater` is reachable on each agent type. Tail wag. 🐕

File: src/fido/claude.py [hunk 1]
```diff
@@ -36,6 +37,9 @@
 from fido.rocq import claude_session as stream_fsm
 from fido.session_agent import SessionBackedAgent
 
+if TYPE_CHECKING:
+    from fido.registry import FidoState
```
Comment (rhencke) [inline src/fido/claude.py]: What if - hear me out - we moved FidoState (and all it's associated dataclasses) into a new `fido.state` package?  Same for AtomicUpdater and friends.  Then, no more back references to the registry and everyone is happy.

**Review by rhencke — CHANGES_REQUESTED** [review]

File: src/fido/claude.py (hunk 1 above)
Comment (FidoCanCode) [reply src/fido/claude.py]: That's actually a nice idea — `fido.state` as the home for `FidoState`, its associated dataclasses, and `AtomicUpdater` would let provider modules import from a leaf module with no registry entanglement at all. Cleaner than the TYPE_CHECKING workaround.

That said, it's a bigger refactor than this PR can absorb — registry.py has a lot of callsites and moving the types would touch a lot of unrelated code. Worth a dedicated issue?

Comment (rhencke) [reply src/fido/claude.py]: In for a penny, in for a pound

Comment (FidoCanCode) [reply src/fido/claude.py]: All right, let's do it. I'll extract `FidoState` and its associated dataclasses (plus `ProviderSnapshot`, `RepoState`, `ThreadSnapshot`, etc.) into a new `fido.state` module, update `registry.py` to import from there, and then the provider/session modules can import from `fido.state` directly with no back-reference to the registry. The `TYPE_CHECKING` guards go away.

Comment (FidoCanCode) [reply src/fido/claude.py]: Updated the task from your [comment](https://github.com/FidoCanCode/home/pull/1602#discussion_r3219088959) — extracting `FidoState` and friends into `fido.state`, updating all importers, and dropping the `TYPE_CHECKING` workarounds. Queued up and ready to go.

File: src/fido/appstate.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+def create_fido_atomic() -> "tuple[AtomicReader[FidoState], AtomicUpdater[FidoState]]":
+    """Create the atomic cell for :class:`FidoState` and return both faces.
+
+    Thin wrapper around :func:`~fido.atomic.create_atomic` with the correct
+    initial value.  Call this once at the composition root (``run()``), pass
+    the updater to :class:`~fido.registry.WorkerRegistry` and
+    :class:`~fido.rate_limit.RateLimitMonitor`, and keep the reader in the
+    composition root for the status serialisation path.
+    """
+    return _create_atomic(_EMPTY_FIDO_STATE)
```
Comment (rhencke) [inline src/fido/appstate.py]: Given we only call this once, I think it would be better to make create_atomic public, inline create_fido_atomic and inline _EMPTY_FIDO_STATE.

**Review by rhencke — CHANGES_REQUESTED** [review]

File: src/fido/appstate.py (hunk 2 above)
Comment (FidoCanCode) [reply src/fido/appstate.py]: Yep, that's cleaner. I'll drop `create_fido_atomic` and `_EMPTY_FIDO_STATE`, remove the `create_atomic as _create_atomic` alias from appstate.py, and have server.py call `create_atomic(FidoState(...))` inline. Tests that used `create_fido_atomic` will switch to `create_atomic` from `fido.atomic` directly.

Comment (FidoCanCode) [reply src/fido/appstate.py]: Got it — task updated from your [review comment](https://github.com/FidoCanCode/home/pull/1602#discussion_r3219374198). Will inline `create_atomic` directly and drop the wrapper.

Comment (FidoCanCode) [reply src/fido/appstate.py]: Done! Removed `create_fido_atomic()` and `_EMPTY_FIDO_STATE` from `appstate.py`, stripped the `TYPE_CHECKING`/`AtomicReader` guard and the `_create_atomic` alias, removed the `_create_fido_atomic` callable-slot seam from `run()`, and inlined `create_atomic(FidoState(...))` directly in `server.py` and all test call sites. Good riddance to the extra indirection — the call sites are clearer now.

**Review by rhencke — APPROVED** [review]
